adr/ADR-43.md

```diff
@@ -0,0 +1,87 @@
+
+| Metadata | Value                     |
+|----------|---------------------------|
+| Date     | 2024-07-11                |
+| Author   | @ripienaar                |
+| Status   | Proposed                  |
+| Tags     | jetstream, client, server |
+
+## Context and motivation 
+
+Streams support a one-size-fits-all approach to message TTL based on the MaxAge setting. This causes any message in the 
+Stream to expire at that age.
+
+There are numerous uses for a per-message version of this limit, some listed below:
+
+ * KV tombstones are a problem in that they forever clog up the buckets with noise, these could have a TTL to make them expire once not useful anymore
+ * Server-applied limits can result in tombstones with a short per message TTL so that consumers can be notified of limits being processed. Useful in KV watch scenarios being notified about TTL removals
+ * KV tombstones can be expanded to also include Limits being applied by the server, with a TTL attached, to facilitate notification of limits in watchers
+ * A stream may have a general MaxAge but some messages may have infinite retention, think a schema or type hints in a KV bucket that is forever while general keys have TTLs
+ * TODO: user use cases
+
+Related issues [#3268](https://github.com/nats-io/nats-server/issues/3268)
+
+## Per-Message TTL
+
+We will allow a message to supply a TTL as a duration in a header called `Nats-TTL` followed by the duration as 
+nanoseconds (TODO: or seconds? I prefer seconds but consistency suggests ns).
+
+The duration will be used by the server to calculate the deadline for removing the message based on its Stream 
+timestamp and the stated duration.
+
+The TTL may not exceed the Stream MaxAge to give administrators a hard cap for the purposes of managing resources.
+
+A TTL of zero or other unparsable value will result in a error (TODO: or ignore? I prefer error) and reported in the 
+Pub Ack.
+
+Any value less than 0 will result in a message being retained forever, even if the stream have a MaxAge set.
+
+## Limit Tombstones
+
+Several scenarios for server-created tombstones can be imagined, the most often requested one though is when MaxAge
+removes last value (ie. the current value) for a Key.
+
+In this case when the server removes a message and the message is the last in the subject it would place a message 
+with a TTL matching the Stream configuration value.  The following headers would be placed:
+
+```
+Nats-Applied-Limit: MaxAge
+Nats-TTL: 1000000000
+```
+
+The `Nats-Limit-Applied` field is there to support future expansion of this feature.
+
+This behaviour is off by default unless opted in on the Stream Configuration.
+
+## Publish Acknowledgements
+
+We could optionally extend the `PubAck` as follows:
+
+```golang
+type PubAck struct {
+	MsgTTL    uint64 `json:"msg_ttl,omitempty"`
+}
+```
+
+This gives clients a chance to confirm, without Stream Info or should the Stream be edited after Info, if the TTL 
```

Comment:
I agree with the reasoning and removing uncertaintity, but this means that because server cannot tell us if it does support given feature, we're having two mechanisms to confirm ack ttl issue - API error (for server supporting `Nats-TTL`) and one for server not supporting it - returning struct without `MsgTTL` field. The concern is that **every** ack will be  bigger by a few bytes, even years in the future, where all streams will support this feature.

Reply:
That then leaves the 2 questions:

 * How do you know if you can use this feature - do we require a stream info always? Even for KV?
 * What if someone updates the stream since your info - kv handles can live a very long time in long running processes

Reply:
This is on the list for issues to work on soon, there's now enough buy in and acknowledgement that this should be addressed at a level that spans all the times we've run into this - so I agree but not for this ADR